### PR TITLE
frozen-account API

### LINKED ISSUE #507: Client API : get a list of frozen accounts
We need to be able to get a list of ‘create-account’ operations that are still frozen.

### PR DESCRIPTION
**Pull request by soonkuk**
### Github Issue
Resolves #507

### CODE REVIEW

File: lib/block/operation.go [hunk 1]
```diff
@@ -152,6 +178,20 @@ func GetBlockOperation(st *storage.LevelDBBackend, hash string) (bo BlockOperati
 	return
 }
 
+func GetBlockOperationCreateFrozen(st *storage.LevelDBBackend, hash string) (bo BlockOperation, err error) {
+	var address string
+	if err = st.Get(GetBlockOperationCreateFrozenKey(hash), &address); err != nil {
+		return
+	}
+	fmt.Println(address)
```
Comment (spikeekips) [inline lib/block/operation.go]: This can be removed :)

File: lib/client/client.go [hunk 2]
```diff
@@ -235,6 +251,34 @@ func (c *Client) StreamAccount(ctx context.Context, id string, cursor *string, h
 	return c.Stream(ctx, url, cursor, handlerFunc)
 }
 
+func (c *Client) StreamFrozenAccountsByLinked(ctx context.Context, id string, cursor *string, handler func(FrozenAccount)) (err error) {
+	url := strings.Replace(UrlAccountFrozenAccounts, "{id}", id, -1)
+	handlerFunc := func(b []byte) (err error) {
+		var v FrozenAccount
+		err = json.Unmarshal(b, &v)
+		if err != nil {
+			return err
```
Comment (spikeekips) [inline lib/client/client.go]: Just `return` will be enough.

File: lib/node/runner/api/account.go [hunk 3]
```diff
@@ -52,3 +56,243 @@ func (api NetworkHandlerAPI) GetAccountHandler(w http.ResponseWriter, r *http.Re
...
+				unfreezing_block_height,
+				unfreezing_opHash,
+				payment_opHash,
+			)
+			txs = append(txs, frozenAccountResource)
+		}
+		closeFunc()
+		return txs
+	}
+	// TODO
```
Comment (spikeekips) [inline lib/node/runner/api/account.go]: Still something to to do?

File: lib/node/runner/api/account.go [hunk 4]
```diff
@@ -52,3 +56,243 @@ func (api NetworkHandlerAPI) GetAccountHandler(w http.ResponseWriter, r *http.Re
...
+		es := NewEventStream(w, r, renderEventStream, DefaultContentType)
+		txs := readFunc()
+		for _, tx := range txs {
+			es.Render(tx)
+		}
+		es.Run(observer.BlockOperationObserver, event)
+		return
+	}
+
+	txs := readFunc() //TODO paging support
```
Comment (spikeekips) [inline lib/node/runner/api/account.go]: Paging and cursor is different?

File: lib/node/runner/api/account.go [hunk 5]
```diff
@@ -52,3 +56,243 @@ func (api NetworkHandlerAPI) GetAccountHandler(w http.ResponseWriter, r *http.Re
...
+				unfreezing_block_height,
+				unfreezing_opHash,
+				payment_opHash,
+			)
+			txs = append(txs, frozenAccountResource)
+		}
+		closeFunc()
+		return txs
+	}
+	// TODO
```
Comment (spikeekips) [inline lib/node/runner/api/account.go]: Same.

**Comment by soonkuk** [comment]
@spikeekips Ok I will remove it.

File: lib/block/account.go [hunk 6]
```diff
@@ -63,6 +63,22 @@ func (b *BlockAccount) Save(st *storage.LevelDBBackend) (err error) {
...
+	if b.Linked != "" {
+		frozenKey := b.NewBlockAccountFrozenKey()
+
+		exists, err = st.Has(frozenKey)
+		if err != nil {
+			return
+		}
+
+		if !exists {
+			err = st.New(frozenKey, b.Address)
```
Comment (anarcher) [inline lib/block/account.go]: Which part check this `err` ?

File: lib/node/runner/api/account.go [hunk 7]
```diff
@@ -52,3 +56,243 @@ func (api NetworkHandlerAPI) GetAccountHandler(w http.ResponseWriter, r *http.Re
 
 	httputils.MustWriteJSON(w, 200, payload)
 }
+
+func (api NetworkHandlerAPI) GetFrozenAccountsByAccountHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	address := vars["id"]
+	options, err := storage.NewDefaultListOptionsFromQuery(r.URL.Query())
+	if err != nil {
+		http.Error(w, errors.ErrorInvalidQueryString.Error(), http.StatusBadRequest)
```
Comment (anarcher) [inline lib/node/runner/api/account.go]:
```suggestion
		httputils.WriteJSONError(w,err)
```
If this `err` is suitable for rendering.

File: lib/node/runner/api/account.go [hunk 8]
```diff
@@ -52,3 +56,243 @@ func (api NetworkHandlerAPI) GetAccountHandler(w http.ResponseWriter, r *http.Re
...
+		var txs []resource.Resource
+		iterFunc, closeFunc := block.GetBlockAccountsByLinked(api.storage, address, options)
+		for {
+			var (
+				create_block_height     uint64
+				create_opHash           string
+				sequenceid              uint64
+				amount                  common.Amount
+				state                   resource.FrozenAccountState
+				unfreezing_block_height uint64
```
Comment (anarcher) [inline lib/node/runner/api/account.go]:
```suggestion
				unfreezingBlockHeight uint64
```

File: lib/node/runner/api/account.go [hunk 9]
```diff
@@ -52,3 +56,243 @@ func (api NetworkHandlerAPI) GetAccountHandler(w http.ResponseWriter, r *http.Re
...
+					payment_opHash = bo.OpHash
+				case operation.TypeUnfreezingRequest:
+					lastblock := block.GetLatestBlock(api.storage)
+					if lastblock.Height-bo.Height >= common.UnfreezingPeriod {
+						state = resource.UnfreezedState
+					} else {
+						state = resource.MeltingState
+					}
+					unfreezing_opHash = bo.OpHash
+					unfreezing_block_height = bo.Height
```
Comment (anarcher) [inline lib/node/runner/api/account.go]:
```suggestion
					unfreezingBlockHeight = bo.Height
```

File: lib/node/runner/api/account.go [hunk 10]
```diff
@@ -52,3 +56,243 @@ func (api NetworkHandlerAPI) GetAccountHandler(w http.ResponseWriter, r *http.Re
...
+			}
+			opCloseFunc()
+			frozenAccountResource := resource.NewFrozenAccount(
+				t,
+				create_block_height,
+				create_opHash,
+				sequenceid,
+				amount,
+				state,
+				unfreezing_block_height,
```
Comment (anarcher) [inline lib/node/runner/api/account.go]:
```suggestion
				unfreezingBlockHeight,
```

File: lib/node/runner/api/account.go (hunk 4 above)
Comment (anarcher) [inline lib/node/runner/api/account.go]:
```suggestion
	txs := readFunc() 
```

File: lib/node/runner/api/resource/frozen_account.go [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+	createdOpHash                string
+	createdSequenceId            uint64
+	initialAmount                common.Amount
+	freezingState                FrozenAccountState
+	unfreezingRequestBlockheight uint64
+	unfreezingRequestOpHash      string
+	paymentOpHash                string
+}
+
+func NewFrozenAccount(ba *block.BlockAccount, create_block_height uint64, create_opHash string, sequenceid uint64, amount common.Amount, state FrozenAccountState, unfreezing_block_height uint64, unfreezing_opHash string, payment_opHash string) *FrozenAccount {
```
Comment (anarcher) [inline lib/node/runner/api/resource/frozen_account.go]: I have a question, Do you have any struct for this fields (sequenceid,create_opHash,state,unfreezing_XXX...) such as `FrozenAccount`?

**Comment by zzim2x** [comment]
I think it should get the results in the order I requested. Currently, it is a random order.

File: lib/node/runner/api/account.go [hunk 12]
```diff
@@ -52,3 +55,249 @@ func (api NetworkHandlerAPI) GetAccountHandler(w http.ResponseWriter, r *http.Re
...
+			state = resource.FrozenState
+			for {
+				bo, hasNext, _ := opIterFunc()
+				switch bo.Type {
+				case operation.TypeUnfreezingRequest:
+					lastblock := block.GetLatestBlock(api.storage)
+					if lastblock.Height-bo.Height >= common.UnfreezingPeriod {
+						state = resource.UnfrozenState
+					} else {
+						unfreezingRemainingBlocks = bo.Height + uint64(241920) - lastblock.Height
```
Comment (zzim2x) [inline lib/node/runner/api/account.go]:
```suggestion
						unfreezingRemainingBlocks = bo.Height + common.UnfreezingPeriod - lastblock.Height
```

File: lib/node/runner/api/resource/frozen_account.go (hunk 11 above)
Comment (soonkuk) [reply lib/node/runner/api/resource/frozen_account.go]: @anarcher I made a struct which have those fields. Thanks.

**Comment by soonkuk** [comment]
> I think it should get the results in the order I requested. Currently, it is a random order.

@zzim2x Yes, I made it sorted.

File: lib/node/runner/api/account.go (hunk 12 above)
Comment (soonkuk) [reply lib/node/runner/api/account.go]: @zzim2x Ok I replaced it.

**Comment by outsideris** [comment]
When will this merge?

File: cmd/sebak/cmd/wallet/unfreezeRequest.go [hunk 13]
```diff
@@ -8,15 +8,14 @@ import (
...
-	"github.com/spf13/cobra"
-
 	cmdcommon "boscoin.io/sebak/cmd/sebak/common"
 	"boscoin.io/sebak/lib/block"
 	"boscoin.io/sebak/lib/common"
 	"boscoin.io/sebak/lib/common/keypair"
 	"boscoin.io/sebak/lib/network"
 	"boscoin.io/sebak/lib/transaction"
 	"boscoin.io/sebak/lib/transaction/operation"
+	"github.com/spf13/cobra"
```
Comment (anarcher) [inline cmd/sebak/cmd/wallet/unfreezeRequest.go]: I think that this import should be upper than `boscoin.io/...` like before :-)

File: lib/block/account.go [hunk 14]
```diff
@@ -63,6 +63,10 @@ func (b *BlockAccount) Save(st *storage.LevelDBBackend) (err error) {
 		createdKey := GetBlockAccountCreatedKey(common.GetUniqueIDFromUUID())
 		err = st.New(createdKey, b.Address)
 	}
+	if err != nil {
+		return
+	}
+
```
Comment (anarcher) [inline lib/block/account.go]: Don't use empty return. In most case, `return err` is more readable IMHO. :-)

File: lib/transaction/operation/operation.go [hunk 15]
```diff
@@ -3,10 +3,9 @@ package operation
 import (
 	"encoding/json"
 
-	"github.com/btcsuite/btcutil/base58"
-
 	"boscoin.io/sebak/lib/common"
 	"boscoin.io/sebak/lib/errors"
+	"github.com/btcsuite/btcutil/base58"
```
Comment (anarcher) [inline lib/transaction/operation/operation.go]: Before this commit, it looks better.

**Pull request merged**